### PR TITLE
SDI-525 Better diagnostic and decreased transaction scope

### PR DESCRIPTION
**Pull request by andymarke**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt [hunk 1]
```diff
@@ -60,20 +60,24 @@ class PrisonerDifferenceService(
       .filter { property -> property.findAnnotations<DiffableProperty>().isNotEmpty() }
       .associate { property -> property.name to property.findAnnotations<DiffableProperty>().first().type }
 
-  @Transactional
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt:63]: No need for transaction to exist while diffs are being generated, hash has already been generated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt [hunk 2]
```diff
@@ -93,7 +97,7 @@ class PrisonerDifferenceService(
       previousPrisonerSnapshot?.also {
         val differences = getDifferencesByCategory(it, prisoner)
         if (differences.isEmpty()) {
-          raiseNoDifferencesTelemetry(offenderBooking.offenderNo)
+          raiseNoDifferencesFoundTelemetry(offenderBooking.offenderNo)
         } else {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt:101]: raise a different event when there should be differences when there isn't compared to there is no differences at all

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt [hunk 3]
```diff
@@ -68,12 +69,17 @@ class PrisonerDifferenceService(
...
     } else {
-      raiseNoDifferencesTelemetry(offenderBooking.offenderNo)
+      raiseNoDifferencesTelemetry(offenderBooking.offenderNo, previousPrisonerSnapshot, prisoner)
     }
   }
 
-  private fun prisonerHasChanged(nomsNumber: String, prisoner: Prisoner): Boolean =
-    prisonerEventHashRepository.upsertPrisonerEventHashIfChanged(nomsNumber, generatePrisonerHash(prisoner), Instant.now()) > 0
+  @Transactional(propagation = Propagation.REQUIRES_NEW)
+  fun prisonerHasChanged(nomsNumber: String, prisoner: Prisoner): Boolean =
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt:77]: attempt at a fix, but not based on much evidence

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**